### PR TITLE
TLS: Clean up ECDSA dependencies

### LINKED ISSUE #7613: TLS: Clean up ECDSA dependencies
This is very similar to https://github.com/Mbed-TLS/mbedtls/issues/7612 but for ECDSA. However, for ECDSA the situation is a bit different since ECDSA signatures can be used in the following cases:

1. With the ECDHE-ECDSA TLS 1.2 key exchange (but _not_ the `ECDH-ECDSA` key exchange which is a historical naming accident).
2. With the TLS 1.3 pure ephemeral key exchange mode.
3. For TLS 1.2 cert-based client authentication, with any key exchange that's using a certificate (not just ECDSA key exchanges, with (EC)DHE-RSA too).

Cases 1 and 2 are similar to ECDH (it's only the list of key exchange (modes) that changes), but case 3 is specific to ECDSA.

Similarly to ECDH, I think we'll want to take advantage of existing macros and define new ones, such as:
- `MBEDTLS_KEY_EXCHANGE_WITH_CLIENT_CERT_ENABLED` = one of TLS 1.2 DHE-RSA, ECDHE-RSA, ECDHE-ECDSA, or TLS 1.3 pure ephemeral - that's when cert-based client auth is possible.
- `MBEDTLS_KEY_EXCHANGE_WITH_CLIENT_ECDSA_CERT_ENABLED` = `MBEDTLS_KEY_EXCHANGE_WITH_CLIENT_CERT_ENABLED && MBEDTLS_PK_CAN_ECDSA_SIGN && MBEDTLS_PK_CAN_ECDSA_VERIFY` - that's when cert-base client auth can be done with ECDSA.
- perhaps 1.2-specific and/or 1.3-specific variants of the above to be used in version-specific code - to be decided while doing this task.

We then want to replace remaining uses of `ECDSA_C`, and some uses of `MBEDTLS_PK_CAN_ECDSA_SOME`, with those new macros.

This task is complete when `grep ECDSA_C library/ssl_*.c library/ssl_*.h include/mbedtls/ssl*.h` finds nothing (and all tests pass unchanged), and in addition use of `MBEDTLS_PK_CAN_ECDSA_SOME` is minimized (ideally, only used to guard the definition of `MBEDTLS_KEY_EXCHANGE_WITH_CLIENT_ECDSA_CERT_ENABLED` and/or other similar macros).

**Conflicts with:** #7612 - likely to change many of the same lines, so better not executed in parallel to avoid merge conflicts.

### PR DESCRIPTION
**Pull request by valeriosetti**
This PR is meant to replace `ECDSA_C` and `PK_CAN_ECDSA_SOME` guards in TLS code with proper key exchanges' related ones.

**Note**: opening as draft because I mostly want to have the CI to find most relevant errors.

Depends on https://github.com/Mbed-TLS/mbedtls/pull/7884
Resolves #7613

## PR checklist

- [x] **changelog** not required as changes are transparent to the end user
- [x] **backport** not required since this is an enhancement
- [x] **tests** not required as only guards are changed, no new code is added

### CODE REVIEW

**Comment by valeriosetti** [comment]
Rebased on `development` after #7884 was merged

**Comment by valeriosetti** [comment]
Sorry I rebased again because I needed the changes introduced by #8011 for testing purposes

**Comment by valeriosetti** [comment]
Here is also the result of `outcome-analysis.sh` script including `ssl-opt` tests:
[oucome-analysis.txt](https://github.com/Mbed-TLS/mbedtls/files/12232600/oucome-analysis.txt)

File: library/ssl_misc.h
Comment (mpg) [inline library/ssl_misc.h:2316]: It seems to me that:
- this macro should be guarding the whole function rather than just the ECDSA block (we don't have to worry about `sig_alg_for_cert_verify` if we're not doing ephemeral, as neither PSK nor PSK-ephemeral use a certificate or a signature).
- the ECDSA block still needs to be guarded by something that means "ECDSA is available". We can either keep using `MBEDTLS_PK_CAN` here, or (perhaps a bit better-looking even if equivalent), define in `ssl_ciphersuites.h` a macro named for example `MBEDTLS_SSL_TLS1_3_KEY_EXCHANGE_MODE_EPHEMERAL_WITH_ECDSA_ENABLED` that is basically `KEY_EXCHANGE_MODE_EPHEMERAL_ENABLED && PK_CAN_ECDSA_SIGN && PK_CAN_ECDSA_VERIFY`.

File: library/ssl_misc.h
Comment (mpg) [inline library/ssl_misc.h:2485]: I don't think that's correct, we can use ECDSA for client auth also with other key exchanges, I think `MBEDTLS_KEY_EXCHANGE_ECDSA_CERT_REQ_ALLOWED_ENABLED` is what we want here.

File: library/ssl_tls.c
Comment (mpg) [inline library/ssl_tls.c:1211]: I think we also want `MBEDTLS_KEY_EXCHANGE_ECDSA_CERT_REQ_ALLOWED_ENABLED` here as this can be used for client auth. (Please correct me if I'm wrong!)

File: library/ssl_tls.c
Comment (mpg) [inline library/ssl_tls.c:1239]: I think we also want `MBEDTLS_KEY_EXCHANGE_ECDSA_CERT_REQ_ALLOWED_ENABLED` here as this can be used for client auth. (Please correct me if I'm wrong!)

File: library/ssl_tls.c
Comment (mpg) [inline library/ssl_tls.c:4984]: I think we also want MBEDTLS_KEY_EXCHANGE_ECDSA_CERT_REQ_ALLOWED_ENABLED here as this can be used for client auth. (Please correct me if I'm wrong!)

Edit: same for all `preset` variables below.

File: library/ssl_tls.c
Comment (mpg) [inline library/ssl_tls.c:5426]: Again, I believe this can be used for client auth, so it needs something more general.

**Review by mpg — CHANGES_REQUESTED** [review]
Only a partial review, I've stopped after a few occurrences where I think there's confusion about the multiple uses of ECDSA, in particular possible use for client auth with key exchanges that don't use ECDSA directly.

The fact that the CI is green while I think a few of the macros are wrong is not good. I think we should add a component that runs `test_suite_ssl` and `tests/ssl-opt.sh -f 'TODO'` in a configuration like full minus `MBEDTLS_KEY_EXCHANGE_ECDHE_ECDSA`, and also the same 1.2-only. If I'm right, this component should fail in the current state of the PR. If not, I'll have some digging to do :)

File: library/ssl_misc.h
Comment (valeriosetti) [reply library/ssl_misc.h:2485]: I apologize for this as well as for the following errors. When replacing previous symbols with `MBEDTLS_KEY_EXCHANGE_ECDHE_ECDSA_ENABLED` I thought that those part of the TLS's code were not related to client authentication, but rather being limited to the key exchange itself.
Obviously that was not correct as add the new components you suggested to `all.sh` I had `ssl-opt.sh` start failing.
I will go through all of those changes and fix them.

Comment (mpg) [reply library/ssl_misc.h:2485]: No worries, this part of the protocol is quite complex. I wasn't 100% sure myself about the things I pointed out, which is why I requested tests, that way we know for sure :)

File: library/ssl_tls.c
Comment (tom-daubney-arm) [reply library/ssl_tls.c:4984]: @valeriosetti I am a bit confused here over what has been implemented here. I was expecting to see `MBEDTLS_KEY_EXCHANGE_ECDSA_CERT_REQ_ALLOWED_ENABLED` here. Did you come to the conclusion that is not warranted or have I missed something?

**Review by tom-daubney-arm — COMMENTED** [review]
I left a comment with a question to hopefully clear up some confusion for myself as I am not sure I am quite understanding fully at this point. Also a small grammatical nit which I have pointed out.

File: library/ssl_tls.c
Comment (valeriosetti) [reply library/ssl_tls.c:4984]: TBH the goal of this PR is not easy also for me, so I might be wrong, but here's my idea behind that.

`MBEDTLS_KEY_EXCHANGE_ECDSA_CERT_REQ_ALLOWED_ENABLED` refers to TLS 1.2 key exchanges, while `MBEDTLS_KEY_EXCHANGE_ECDSA_CERT_REQ_ANY_ALLOWED_ENABLED` would be more appropriate for referring to both TLS 1.2 and 1.3. However, since the symbols defined in this array concern TLS 1.3, I thought that just keeping the TLS 1.3 part of `MBEDTLS_KEY_EXCHANGE_ECDSA_CERT_REQ_ANY_ALLOWED_ENABLED` (i.e. `MBEDTLS_SSL_TLS1_3_KEY_EXCHANGE_MODE_EPHEMERAL_ENABLED`) could be enough.

The CI seems to agree with that, but as I'm not an expert at all of this, so my conclusion might be wrong. Any suggestion is accepted on this :)

Comment (tom-daubney-arm) [reply library/ssl_tls.c:4984]: Yeah, what your saying does make sense to me. I am also not an expert and this stuff is confusing. I will have another think about it and re-review.

**Comment by valeriosetti** [comment]
The CI was almost entirely green, but I had to force push the last commit  (without any code change) because it was stuck on 1 task for 3 days. I think the PR is still ready for review since this last commit only fixes a typo and the preceding commit got a fully green CI

**Review by tom-daubney-arm — APPROVED** [review]
At this point the PR LGTM.

The fact that a new testing component was added which initially failed and then passed after corrections had been applied is reassuring as it implies there are not any test gaps of major concern here.

File: include/mbedtls/ssl_ciphersuites.h
Comment (xkqian) [inline include/mbedtls/ssl_ciphersuites.h:303]: Seems we have two separate macro guards for the cert-based `key exchange `and `client authentication `in TLS1.2, but for TLS1.3 we only one guard to protect them.   Is "guard them separately" better?

**Review by xkqian — COMMENTED** [review]
It looks quite good to me. 
I just leave some minor questions.
We have two separate macro guards for the cert-based `key exchange `and `client authentication `in TLS1.2, but for TLS1.3 we only one guard to protect them.  
Which way is better? Should we sync them?

File: include/mbedtls/ssl_ciphersuites.h
Comment (mpg) [reply include/mbedtls/ssl_ciphersuites.h:303]: Sorry, can you clarify which are the two macros fro 1.2 and the single macro for 1.3 that you are referring to?

The protocols are different regarding how authentication works, and I think having a single macro for 1.3 is correct, but now that you mention it, perhaps a single macro is enough for 1.2 as well. I you could kindly indicate the names of the macros you're referring to for each protocol, I'll double-check whether we need exactly those, or perhaps one more or one less.

Comment (xkqian) [reply include/mbedtls/ssl_ciphersuites.h:303]: > can you clarify which are the two macros fro 1.2 and the single macro for 1.3 that you are referring to?

For TLS1.2, I just see MBEDTLS_KEY_EXCHANGE_WITH_CERT_ENABLED for the `cert based key exchange`, and MBEDTLS_KEY_EXCHANGE_ECDSA_CERT_REQ_ALLOWED_ENABLED) for the `cert based client authentication`.  And for TLS1.3, we always use MBEDTLS_SSL_TLS1_3_KEY_EXCHANGE_MODE_EPHEMERAL_ENABLED to guard both the cert based `key exchange `and `client authentication`.

Comment (mpg) [reply include/mbedtls/ssl_ciphersuites.h:303]: Thanks for the precisions! Looking at the macros are defined, I think this is correct and reflects differences the protocols:
- in TLS 1.3, all key exchange modes that allow a server certificate also allow a client certificate ("all" is actually just one mode: Ephemeral);

File: library/ssl_tls.c
Comment (mpg) [reply library/ssl_tls.c:4984]: > However, since the symbols defined in this array concern TLS 1.3

Ah, but that's the tricky part: even though those symbols have TLS 1.3 in their name, they're not just for 1.3: actually `MBEDTLS_TLS1_3_SIG_ECDSA_SECP256R1_SHA256` can also be spelt `MBEDTLS_SSL_TLS12_SIG_AND_HASH_ALG(MBEDTLS_SSL_SIG_ECDSA, MBEDTLS_SSL_HASH_SHA512)` and then it suddently doesn't looks that specific to TLS 1.3 any more :) (It's not just a numerical coincidence, it was done on purpose, see [RFC 8446 section 4.2.3](https://datatracker.ietf.org/doc/html/rfc8446#section-4.2.3), last-but-one paragraph: "ECDSA signature schemes align with TLS 1.2's ECDSA hash/signature pairs."

Coming back to our code, the comment just above this says that `ssl_preset` "is for TLS 1.3 only or hybrid TLS 1.3/1.2 handshakes", so for algorithms that are common with TLS 1.2, we need to use a `ANY` macro, not a `TLS1_3` macro.

Also, for TLS 1.3, we shouldn't remove the `PK_CAN_ECDSA` condition: we we can't do ECDSA, we shouldn't advertise it.

File: include/mbedtls/ssl_ciphersuites.h
Comment (mpg) [reply include/mbedtls/ssl_ciphersuites.h:303]: I'm adding a comment to clarify.

**Comment by mpg** [comment]
I'm adopting this. I've pushed an update that should address @xkqian's feedback, as well the point from my initial review where @tom-cosgrove-arm expressed doubts about the resolution. I've also double-checked that all other points from my initial review had been addressed.

So, I believe this is ready for re-review.

**Review by mpg — APPROVED** [review]

**Comment by mpg** [comment]
Aw, I misclicked - I meant to dismiss my initial review, not approve. Please ignore my approval, it can't count as I've pushed commits of my own, we still need two re-reviews. Sorry for the confusion.

**Comment by tom-cosgrove-arm** [comment]
@mpg I've "requested" a review from you, to dismiss your mis-click review!

**Review by tom-daubney-arm — APPROVED** [review]
Thanks for the changes. LGTM.

**Review by tom-daubney-arm — APPROVED** [review]
Not sure why the previous review was dismissed but I am approving again.

**Review by paul-elliott-arm — APPROVED** [review]
LGTM

**Comment by gilles-peskine-arm** [comment]
The [failure in the merge queue](https://mbedtls.trustedfirmware.org/job/mbed-tls-pr-head/job/gh-readonly-queue%252Fdevelopment%252Fpr-7921-452beb9076d2e404ebaae9a4944baa806cdb09b5/1/display/redirect) is obviously not related to this pull request. I believe that it's a rare race condition in parallel builds. I'm going to file an issue with my analysis. I've re-triggered the merge.

**Review by xkqian — COMMENTED** [review]
LGTM

File: library/ssl_misc.h
Comment (mpg) [reply library/ssl_misc.h:2316]: Note: I incorrectly marked this conversation as resolved. This change is incorrect for two reasons:
- the new guard redundant, this whole region of code is already guarded by it (line 2295 above);
- we still need to check if ECDSA is available, so we shoudn't remove `MBEDTLS_PK_CAN_ECDSA_SOME` - unless we replace it with something that somehow also expresses "ECDSA is available".

This was fixed when rebasing #8075 after this PR was merged, see https://github.com/Mbed-TLS/mbedtls/pull/8075#discussion_r1333516744

File: tests/ssl-opt.sh
Comment (ronald-cron-arm) [inline tests/ssl-opt.sh:1634]: As part of the review of #8526, I am investigating in the configuration of all.sh component test_tls13_only_psk + MBEDTLS_SSL_RECORD_SIZE_LIMIT why "Record Size Limit: TLS 1.3: Server-side parsing and debug output" is run but "Record Size Limit: TLS 1.3: Client-side parsing and debug output" is skipped. I ended up here and I do not understand this change. With this commit, given the checks in `use_ext_tool_without_ecdh_support`, it basically means that as soon as we use GnuTLS in one test TLS_VERSION is going to be TLS12 and thus that the dependencies checks for TLS 1.3 tests below using "detect_required_features" are going to be wrong.
@valeriosetti @mpg @paul-elliott-arm @tom-daubney-arm what am I missing here?

Comment (valeriosetti) [reply tests/ssl-opt.sh:1634]: Question: I checked out at #8526 (current status) and then I added `scripts/config.py set   MBEDTLS_SSL_RECORD_SIZE_LIMIT` to `component_test_tls13_only_psk`. However in this configuration I get both `"Record Size Limit: TLS 1.3: (Client|Server)-side parsing and debug output"` being skipped. I'm running this in Docker with Ubuntu 16.04 so GnuTLS versions should be fine. Am I missing something?

A part from this, if in the same Docker container I issue the command `$GNUTLS_CLI -l` to list all available ciphers in GnuTLS, I don't see any TLS1.3 listed there. TBH I don't remember exactly the reason for which I added the check you mentioned, but this sounds like a possible reason. Wdyt?

Comment (mpg) [reply tests/ssl-opt.sh:1634]: I don't think you're missing anything and I can no longer understand why things are the way they are, so either I missed an issue in reviewing this, or we really should have added more comments - which would also be something I missed when reviewing this...

Specifically:
- I think `use_ext_tool_without_ecdh_support` is misleading since is document talks about _static_ ECDH while the name suggests it's about all forms of ECDH. (I know ephemeral is called ECDHE in TLS, but still, "ECDH" does not clearly convey "static" to me.) Actually a better name for the current implementation would be `peer_is_not_mbedtls()` or something. (Similarly for the variable holding the result.)
- I have no idea why we always assume TLS 1.2 when using OpenSSL or GnuTLS as the peer. This might be justified in some particular cases but does not seem correct in general.

Comment (ronald-cron-arm) [reply tests/ssl-opt.sh:1634]: Thanks for looking at this. #8526 has evolved since my previous comment thus now (as of 55dfa2c06) if you remove the dependency "requires_config_enabled MBEDTLS_SSL_TLS1_3_KEY_EXCHANGE_MODE_EPHEMERAL_ENABLED" you restore the issue. I've just checked it again and the "Server-side" test run and fail (as expected because we need the ephemeral key exchange mode and it is not enabled) while the "Client-side" is skipped where I would expect it to run and fail as well.   
Otherwise, the tests use G_NEXT_CLI/SRV. On my machine, the tests are run with GnuTLS 3.7.2 (version that supports TLS 1.3) and in tests/docker/bionic/Dockerfile I can see:
```
ENV GNUTLS_NEXT_CLI=/usr/local/gnutls-3.7.2/bin/gnutls-cli                      
ENV GNUTLS_NEXT_SERV=/usr/local/gnutls-3.7.2/bin/gnutls-serv
```
Thus you should be able to reproduce in Docker.

Comment (gilles-peskine-arm) [reply tests/ssl-opt.sh:1634]: It's quite possible that I made some mistake, perhaps I wrote something for TLS ≤1.2 and didn't realize that it didn't work the way I intended for 1.3?

I'm not sure what `use_ext_tool_without_ecdh_support` is about. AFAIR there are some versions of OpenSSL/GnuTLS/both that only support ECDHE and not ECDH, but that wouldn't affect TLS 1.3, only ≤1.2.

Comment (mpg) [reply tests/ssl-opt.sh:1634]: >I have no idea why we always assume TLS 1.2 when using OpenSSL or GnuTLS as the peer. This might be justified in some particular cases but does not seem correct in general.

What I suspect now, but can't be sure, is that this was written (and reviewed) in a context where we were focusing on things failing with static-ECDH tests, and we simply forgot the more general context.

Now, I think the only way for a test case to require static-ECDH is if an `ECDH_xxx` ciphersuite is forced (otherwise ephemeral ECDH will be available and preferred). So if we somehow need special handling of static-ECDH I think we should look at the `force_ciphersuite` command line option to determine if we're using static-ECDH or not.

But I'm not sure we should need special handling of static-ECDH: presumably the test cases that force a static-ECDH ciphersuite don't try to use a peer that doesn't support static-ECDH.

Also, when guessing which TLS version will be used, we should have a look at `force_ciphersuite` too: if the ciphersuite name contains `WITH` then it's TLS 1.2 otherwise TLS 1.3.

Comment (mpg) [reply tests/ssl-opt.sh:1634]: @ronald-cron-arm I think you should feel free to modify this in any way that looks logical to you, as long as it keeps the CI passing and doesn't result in more test being skipped in the default or full config.

That is, I don't think there's any clever reason that you might be missing for why things are this way. We probably just made a mistake and introduced a bug while trying to fix another one. This evolved quite organically (mostly in response to CI failures) so it's quite likely that we just lost track of the big picture at some point.

I guess it wouldn't hurt to take a step back and consider how we'd like all this auto-detection to work, and possibly restructure it a bit. (Not sure if now's the right time though.)

Comment (gilles-peskine-arm) [reply tests/ssl-opt.sh:1634]: Filed as https://github.com/Mbed-TLS/mbedtls/issues/9560.

(After spending hours trying to understand what was going on :frowning_face: Please file issues when you know something is wrong!)

Comment (mpg) [reply tests/ssl-opt.sh:1634]: > (After spending hours trying to understand what was going on ☹️ Please file issues when you know something is wrong!)

Sorry, I'll pay more attention in the future.

**Pull request merged**
